test/unit/context/test-node-discover.js

```diff
@@ -0,0 +1,958 @@
+    }
+  }
+
+  describe('ContextNode.get', () => {
+    it('should create an element node', () => {
+      const el = doc.createElement('div');
+      const cn = ContextNode.get(el);
+      expect(cn.node).to.equal(el);
+      // Parent always starts as null.
+      expectContext(cn, {
+        parent: null,
+        isRoot: false,
+        root: null,
+        children: [],
+        discoverable: true,
+      });
+    });
+
+    it('should create a fragment node', () => {
+      const frag = doc.createDocumentFragment();
+      const cn = ContextNode.get(frag);
+      expect(cn.node).to.equal(frag);
+      // Parent always starts as null.
+      expectContext(cn, {
+        parent: null,
+        isRoot: false,
+        root: null,
+        children: [],
+        discoverable: true,
+      });
+    });
+
+    it('should create a document node', () => {
+      const cn = ContextNode.get(doc);
+      expect(cn.node).to.equal(doc);
+      // Parent always starts as null.
+      expectContext(cn, {
+        parent: null,
+        isRoot: true,
+        root: doc,
+        children: [],
+        discoverable: false,
+      });
+    });
+
+    it('should create a node only once', () => {
+      const el = doc.createElement('div');
+      const frag = doc.createDocumentFragment();
+      const cn1 = ContextNode.get(el);
+      const cn2 = ContextNode.get(frag);
+      const cn3 = ContextNode.get(doc);
+      expect(cn1).to.equal(ContextNode.get(el));
+      expect(cn2).to.equal(ContextNode.get(frag));
+      expect(cn3).to.equal(ContextNode.get(doc));
+    });
+  });
+
+  describe('ContextNode.closest', () => {
+    let element;
+    let grandparent;
+
+    beforeEach(() => {
+      element = el('T-1-1-1');
+      grandparent = el('T-1');
+    });
+
+    it('should find itself', () => {
+      const elementContext = ContextNode.get(element);
+      expect(ContextNode.closest(element)).to.equal(elementContext);
+    });
+
+    it('should skip itself', () => {
+      ContextNode.get(element);
+      const grandparentContext = ContextNode.get(grandparent);
+      expect(ContextNode.closest(element, false)).to.equal(grandparentContext);
+      expect(ContextNode.closest(grandparent, false)).to.be.null;
+    });
+
+    it('should go up DOM tree', () => {
+      const grandparentContext = ContextNode.get(grandparent);
+      expect(ContextNode.closest(element)).to.equal(grandparentContext);
+    });
+
+    it('should auto-create root when connected', () => {
+      doc.body.appendChild(tree);
+      expect(ContextNode.closest(element)).to.equal(ContextNode.get(doc));
+      expectContext(ContextNode.get(doc), {isRoot: true});
+    });
+
+    it('should auto-create root on a fragment', () => {
+      const frag = doc.createDocumentFragment();
+      frag.appendChild(tree);
+      expect(ContextNode.closest(element)).to.equal(ContextNode.get(frag));
+      expectContext(ContextNode.get(frag), {
+        isRoot: false,
+        root: null,
```

Comment:
why is `root` null here considering it actually is the root?

Reply:
Fragments are not roots by default b/c fragments are ambiguous. It could be a root, or could just be part of a subtree. The party that creates a fragment is responsible to either:

1. Keep it w/o root until later. E.g. I can create a fragment, build lots of DOM, and only later attach it to the live DOM. You can think of this case as a `<template>` which contains inert DOM. Actually, template could be such a fragment.
2. Attach the fragment to an existing subtree by calling `setParent()`. E.g. a shadow root of a component would normally be just part of the main tree.
3. Make fragment itself a root by calling `setIsRoot()`. E.g. a shadow root that contains a subdocument (shadow doc) would normally be made the whole root itself. A similar case is with amp-next-page.

The (1) is simply a question of doing as little work as possible until the node is attached. The distinction between (2) and (3) is driven mainly by the question when can values propagate between the subtrees and when they can't. Propagating a service into a wrong subtree could have a lot of risks. And so I opted not to make an automatic decision and leave it up to the fragment creator to decide.